ENH: integrate.solve_ivp: critical time support

MatthewFlamm opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

Likely closes #14246

#### What does this implement/fix?
This PR implements critical time support for `solve_ivp`. The detailed scipy roadmap currently says

> Complete the feature set of solve_ivp, including all functionality of the odeint interface.

One of the features of `odeint` is to provide a `tcrit` input parameter as

>tcrit: ndarray, optional
>> Vector of critical points (e.g., singularities) where integration care should be taken.

In `odeint` this means that the solution should integrate to exactly those points, then proceed further.


#### Additional information
Currently the end point of simulation is being used as a critical time point in the LSODA solver.  This PR modifies the logic to progressively step through a series of critical points.  A test is added that ensures this point is hit exactly.

~This feature is likely possible to implement for all the solve_ivp solvers, e.g. RK could limit the next time step similar to the `t_bound` check below.  While I'm willing to implement for all solvers, I wanted to get feedback on this approach before spending that time.~

This section has now been modified for all solvers.

https://github.com/scipy/scipy/blob/fbb6ff839cfdf02304a3bd21d46a1189c01832e9/scipy/integrate/_ivp/rk.py#L143-L144

#### Areas of feedback
- Should the monotonicity of tcrit be checked explicitly or assumed to be correct (current behavior)?
- ~The current code silently excludes the critical times outside the interval of integration, should it raise an error or warning instead?~
- ~Implement for all other solvers?~
  - ~If so, move common tcrit logic to ODESolver~


### TODO

These items depend on the feedback above.

- [ ] Add tests for tcrit sequences that are non-monotonically increasing or decreasing ~or fall outside t0 or tbound.~
- [x] Add tests for the tcrit attribute on LSODA class.
- [x] use for other solvers

MatthewFlamm committed:
ENH: Add tcrit to LSODA

MatthewFlamm commented on scipy/integrate/_ivp/lsoda.py:
```diff
@@ -207,6 +227,38 @@ def _dense_output_impl(self):
+        self.tcrit = np.append(tcrit, t_bound)
+
+    def _find_next_tcrit(self):
+        if self.tcrit.size == 0:
```
From a performance perspective, these extra if statements will give extra overhead.  It would be better if the minimum length of `tcrit` is `1`, i.e. `t_bound` is always there.  Then we only need to handle this extra logic when a critical point is hit, which is infrequent.

MatthewFlamm replied:
changed in https://github.com/scipy/scipy/pull/23385/commits/a96c26db331c01aba1a15f433865aac951053983

MatthewFlamm committed:
mark lsoda test as nonthreadsafe

MatthewFlamm committed:
more performant check tcrit check

MatthewFlamm committed:
Merge branch 'main' into feat/lsoda-solve_ivp-tcrit

MatthewFlamm committed:
Merge branch 'main' into feat/lsoda-solve_ivp-tcrit

MatthewFlamm committed:
Merge branch 'main' into feat/lsoda-solve_ivp-tcrit

ilayn commented:
Ah dangit, I hope I did not create too much hassle for your work here with the rewrite. I should have checked before, apologies.

MatthewFlamm committed:
LSODA now thread-safe

MatthewFlamm committed:
make documentation reference t_span

MatthewFlamm committed:
tests for backward time; fix tcrit for backwards

MatthewFlamm committed:
error with out of bounds

MatthewFlamm committed:
move functionality to base class

MatthewFlamm committed:
implement for BDF

MatthewFlamm committed:
implement for Radau

MatthewFlamm committed:
implement for RK

MatthewFlamm committed:
documentation

MatthewFlamm committed:
remove unused var

MatthewFlamm commented:
This PR was extended to add the functionality to all solvers in `solve_ivp`.  I updated the PR description and title accordingly.

There is a lot of common functionality in the Python solvers that could be considered for refactoring, but I kept it simple for the PR and implemented the same logic in each.

MatthewFlamm committed:
remove reference to LSODA in docs

MatthewFlamm committed:
fix whitespace

MatthewFlamm committed:
Merge branch 'main' into feat/lsoda-solve_ivp-tcrit

MatthewFlamm commented on scipy/integrate/_ivp/base.py:
```diff
@@ -101,6 +101,15 @@ class OdeSolver:
+        ascending or descending in the same manner as the direction
+        between ``t0`` and ``t_bound``.
+
+        .. versionadded:: 1.17.0
```
```suggestion
        .. versionadded:: 1.18.0
```

MatthewFlamm commented on scipy/integrate/_ivp/bdf.py:
```diff
@@ -155,6 +155,15 @@ class BDF(OdeSolver):
+        ascending or descending in the same manner as the direction
+        between ``t0`` and ``t_bound``.
+
+        .. versionadded:: 1.17.0
```
```suggestion
        .. versionadded:: 1.18.0
```

MatthewFlamm commented on scipy/integrate/_ivp/ivp.py:
```diff
@@ -360,6 +360,15 @@ def solve_ivp(fun, t_span, y0, method='RK45', t_eval=None, dense_output=False,
+        ascending or descending in the same manner as the direction
+        defined by ``t_span``.
+
+        .. versionadded:: 1.17.0
```
```suggestion
        .. versionadded:: 1.18.0
```

MatthewFlamm commented on scipy/integrate/_ivp/radau.py:
```diff
@@ -258,6 +258,15 @@ class Radau(OdeSolver):
+        ascending or descending in the same manner as the direction
+        between ``t0`` and ``t_bound``.
+
+        .. versionadded:: 1.17.0
```
```suggestion
        .. versionadded:: 1.18.0
```

MatthewFlamm commented on scipy/integrate/_ivp/rk.py:
```diff
@@ -339,6 +341,15 @@ class RK45(RungeKutta):
+        ascending or descending in the same manner as the direction
+        between ``t0`` and ``t_bound``.
+
+        .. versionadded:: 1.17.0
```
```suggestion
        .. versionadded:: 1.18.0
```

MatthewFlamm committed:
update versionadded

MatthewFlamm committed:
Merge branch 'main' into feat/lsoda-solve_ivp-tcrit

MatthewFlamm committed:
Merge branch 'main' into feat/lsoda-solve_ivp-tcrit

MatthewFlamm committed:
Merge branch 'main' into feat/lsoda-solve_ivp-tcrit

MatthewFlamm committed:
Remove extraneous blank lines.

Removed unnecessary blank lines in the _ivp/base.py file.

MatthewFlamm committed:
Merge branch 'main' into feat/lsoda-solve_ivp-tcrit

MatthewFlamm committed:
Remove duplicated lines from merge conflicts

Co-authored-by: MatthewFlamm <39341281+MatthewFlamm@users.noreply.github.com>
